policy-controller/k8s/index/src/index.rs

```diff
@@ -0,0 +1,712 @@
+                entry.insert(PodPortServer { name: None, tx, rx });
+            }
+
+            Entry::Occupied(mut entry) => {
+                let ps = entry.get_mut();
+
+                // Avoid sending redundant updates.
+                if *ps.rx.borrow() == server {
+                    return;
+                }
+
+                ps.name = None;
+                ps.tx.send(server).expect("a receiver is held by the index");
+            }
+        }
+    }
+
+    /// Enumerates ports.
+    ///
+    /// A named port may refer to an arbitrary number of port numbers.
+    fn select_ports(&mut self, port_ref: &Port) -> Vec<u16> {
+        match port_ref {
+            Port::Number(p) => Some(*p).into_iter().collect(),
+            Port::Name(name) => self
+                .port_names
+                .get(name)
+                .into_iter()
+                .flatten()
+                .cloned()
+                .collect(),
+        }
+    }
+
+    fn port_server_or_default(&mut self, port: u16, config: &ClusterInfo) -> &mut PodPortServer {
+        match self.port_servers.entry(port) {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => {
+                let (tx, rx) = watch::channel(Self::default_inbound_server(
+                    port,
+                    &self.meta.settings,
+                    config,
+                ));
+                entry.insert(PodPortServer { name: None, tx, rx })
+            }
+        }
+    }
+
+    fn default_inbound_server(
+        port: u16,
+        settings: &pod::Settings,
+        config: &ClusterInfo,
+    ) -> InboundServer {
+        let protocol = if settings.opaque_ports.contains(&port) {
+            ProxyProtocol::Opaque
+        } else {
+            ProxyProtocol::Detect {
+                timeout: config.default_detect_timeout,
+            }
+        };
+
+        let mut policy = settings.default_policy.unwrap_or(config.default_policy);
+        if settings.require_id_ports.contains(&port) {
+            if let DefaultPolicy::Allow {
+                ref mut authenticated_only,
+                ..
+            } = policy
+            {
+                *authenticated_only = true;
+            }
+        }
+
+        let mut authorizations = HashMap::default();
+        if let DefaultPolicy::Allow {
+            authenticated_only,
+            cluster_only,
+        } = policy
+        {
+            let authentication = if authenticated_only {
+                ClientAuthentication::TlsAuthenticated(vec![IdentityMatch::Suffix(vec![])])
+            } else {
+                ClientAuthentication::Unauthenticated
+            };
+            let networks = if cluster_only {
+                config.networks.iter().copied().map(Into::into).collect()
+            } else {
+                vec![
+                    "0.0.0.0/0".parse::<IpNet>().unwrap().into(),
+                    "::/0".parse::<IpNet>().unwrap().into(),
```

Comment:
nit, TIOLI: would maybe be nice if there was a way to construct these without parsing them from strings every time?